### PR TITLE
Rutefig/reg 654 zk regex fix 1 character capture group extraction fails in

### PR DESCRIPTION
**Pull request by rutefig**
## Summary

This PR fixes a critical bug in capture group extraction for 1-character captures in both Circom and Noir circuits. Previously, single-byte capture groups would fail to extract correctly, causing ZK proofs to fail for patterns with short captures like `from:x\r\n` where `x` is the captured value.

### Root Cause

The backward sweep in `CaptureSubstring` (Circom) and `build_capture_start_end_mask` (Noir) used `captureStarts[i] == 0` to detect capture "continuation" bytes. However, single-byte captures only have a start marker (`1`), with no continuation bytes (`0`). This caused the backward sweep to skip 1-char captures entirely.

### The Fix

Changed the backward sweep logic in both Circom and Noir to use `isCapture[i]` directly instead of checking for continuation bytes. Group membership (not byte position within the group) determines the capture mask.

**Circom (`regex_helpers.circom`):**
```diff
- isCaptureEnd[i] <== IsEqual()([captureStarts[i], 0]);
- isValidCaptureEndTemp[i] <== AND()(isCapture[i], isCaptureEnd[i]);
+ isValidCaptureEndTemp[i] <== isCapture[i];
```

**Noir (`captures.nr`):**
```diff
- let is_capture_end_temp = (capture_starts[j] == 0) as Field;
- let is_capture_end_valid_temp = (is_capture[j] * is_capture_end_temp == 1) as Field;
+ let is_capture_end_valid_temp = is_capture[j];
```

### Changes Included

- **Core fix**: Updated backward sweep logic in both circuit implementations
- **Documentation**: Added comprehensive comments explaining the two-sweep algorithm
- **Compiler note**: Documented BTreeSet ordering dependency for capture group handling
- **Tests**: Added `mixed_capture_length` regex pattern with tests for:
  - Single char in first capture, multi-char in second (`a_hello`)
  - Both captures being 1 character (`x_y`)
  - Multi-char variations (`a_bcdef`)
- **Regenerated templates**: All Circom and Noir circuit templates regenerated with the fix

## Test Plan

- [x] All existing tests pass (`bun test` and `nargo test`)
- [x] New short capture tests verify 1-char extraction works correctly
- [x] Tests cover edge cases: single-char in both capture groups, mixed lengths

### Manual Verification

The fix can be verified by running:

```bash
bun run test:circom
cd noir && nargo test
```

All tests should pass, including the new `short_capture.test.ts` tests that specifically target 1-character capture group extraction.

## Linear Issue

[REG-654](https://linear.app/zk-email/issue/REG-654): ZK Regex fix 1-character capture group extraction fails in Circom and Noir

## Summary by CodeRabbit

* **New Features**
  * Added support for regex patterns with mixed capture groups, enabling extraction of multiple matched segments with configurable length constraints.
  * Enhanced regex matching engine with improved capture group handling and validation.

* **Tests**
  * Added comprehensive test suite with multiple test circuits for pattern matching scenarios.
  * Extended test fixtures covering various capture group configurations and input patterns.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by rutefig** [comment]
@coderabbitai review

**Comment by rutefig** [comment]
@BornPsych I believe all of these comments from @coderabbitai are from template generation so not relevant for this PR but we can track this and check later on

File: circom/circuits/regex_helpers.circom [hunk 1]
```diff
@@ -118,49 +133,82 @@ template CaptureSubstring(maxBytes, maxSubstringBytes, captureId) {
...
-        isCaptureEnd[i] <== IsEqual()([captureStarts[i], 0]);
-        isValidCaptureEndTemp[i] <== AND()(isCapture[i], isCaptureEnd[i]);
+        // Is this a byte of *our* capture group?
+        isValidCaptureEndTemp[i] <== isCapture[i];
         if (i == maxBytes - 1) {
+            // Base case: no positions to the right to inherit from
             isValidCaptureEnd[i] <== isValidCaptureEndTemp[i];
         } else {
+            // Cumulative OR: propagates rightmost capture content leftward
             isValidCaptureEnd[i] <== OR()(isValidCaptureEndTemp[i], isValidCaptureEnd[i+1]);
```
Comment (BornPsych) [inline circom/circuits/regex_helpers.circom]: `isValidCaptureEndTemp[i] <== isCapture[i];`

This creates an extra signal array that just copies isCapture[i]. Each assignment is a constraint. we can also remove the declaration from the top

```suggestion
if (i == maxBytes - 1) {
            // Base case: no positions to the right to inherit from
isValidCaptureEnd[i] <== isCapture[i];
} else {
            // Cumulative OR: propagates rightmost capture content leftward
isValidCaptureEnd[i] <== OR()(isCapture[i], isValidCaptureEnd[i+1]);
```

File: circom/circuits/regex_helpers.circom [hunk 2]
```diff
@@ -118,49 +133,82 @@ template CaptureSubstring(maxBytes, maxSubstringBytes, captureId) {
...
 
+    // Phase 1: Forward sweep (left → right)
+    // Builds isValidCaptureStart[i] which answers:
+    //   "At position i, have we already seen the start of our capture group?"
+    // Once the start is found, the flag stays 1 for all subsequent positions.
     for (var i = 0; i < maxBytes; i++) {
+        // Is this byte marked as the start of any capture group?
         isCaptureStart[i] <== IsEqual()([captureStarts[i], 1]);
+        // Is this the start of *our* capture group? (belongs to our group AND is a start marker)
         isValidCaptureStartTemp[i] <== AND()(isCapture[i], isCaptureStart[i]);
```
Comment (BornPsych) [inline circom/circuits/regex_helpers.circom:166]:
```suggestion
       isValidCaptureStartTemp[i] <== isCapture[i] * captureStarts[i];
```

Since both  isCapture[i]  and  captureStarts[i]  are binary (0 or 1), and it saves 2 constraints

**Review by BornPsych — APPROVED** [review]

File: circom/circuits/regex_helpers.circom (hunk 1 above)
Comment (rutefig) [reply circom/circuits/regex_helpers.circom]: Yes you're right, the assertion is redundant and the `isValidCaptureEndTemp` signal is not needed. Removing it, thanks for noticing this

File: circom/circuits/regex_helpers.circom (hunk 2 above)
Comment (rutefig) [reply circom/circuits/regex_helpers.circom:166]: yes that is correct, but my concern with that approach is that if we consider this template as a standalone using `* ` without binary assertion (which we don't have it here, we currrently have this assertion on the parent template) can cause issues since we are assuming both signals are binary. When using the `Equal` and `AND` gates even tho we are creating more constraints we don't need to assume they are binary, it will work anyways

Comment (BornPsych) [reply circom/circuits/regex_helpers.circom:166]: DO you think by any chance we can have captureStarts and isCapture, non-binary, and if we do these changes we can save  it saves 2 constraints per bye which means, for email of size 4096, we can save almost 8192 constraints which is huge

Comment (rutefig) [reply circom/circuits/regex_helpers.circom:166]: After discussion we decided to move this to another [issue](https://linear.app/zk-email/issue/REG-658) for further investigation
